File: cse241-object-oriented-programming/studies/cpp/ch15-Polymorphism-and-Virtual-Functions/employee/hourlyEmployee.cpp
#include <iostream>
#include <string>
#include "hourlyEmployee.h"
using std::cout;
using std::string;

namespace GTUEmployee {
    HourlyEmployee::HourlyEmployee (const string & firstName, const string & lastName, const string ssn, double wage, double hours) 
    : Employee(firstName, lastName, ssn) {
        setWage(wage);
        setHours(hours);
    }

    void HourlyEmployee::setWage (double wage) {
        _wage = (wage < 0.0) ? 0.0 : wage;
    } 
    void HourlyEmployee::setHours (double hours) {
        _hours = (hours < 0.0 || hours > 168.0) ? 0.0 : hours;

    } 

    double HourlyEmployee::getWage () const {return _wage;} 
    double HourlyEmployee::getHours () const {return _hours;} 

    double HourlyEmployee::earnings () const {
        if (getHours() <= 40.0) // no overtime
            return getWage() * getHours();
        else
            return 40 * getWage() + ( ( getHours() - 40 ) * getWage() * 1.5 );
    }

    void HourlyEmployee::print () const {
        cout << "hourly employee: ";
        Employee::print(); // code reuse
        cout << "\nhourly wage: " << getWage() << "; hours worked: " << getHours();    
    }
}

File: cse241-object-oriented-programming/studies/java/inheritance/employee/HourlyEmployee.java
package employee;

public class HourlyEmployee extends Employee {
    private double _hourlyWage;   // wage per hour 
    private double _hoursWorked;  // hours wored for week

    public HourlyEmployee (String fname, String lname, String ssn, double hourlyWage, double hoursWorked) {
        super(fname, lname, ssn);
        setHourlyWage(hourlyWage);
        setHoursWorked(hourlyWage);
    }

    public void setHourlyWage (double hourlyWage) {_hourlyWage = hourlyWage < 0.0 ? 0.0 : hourlyWage;}
 
    public void setHoursWorked (double hoursWorked) {_hoursWorked = hoursWorked < 0.0 ? 0.0 : hoursWorked;}

    public double hourlyWage () {return _hourlyWage;}
 
    public double hoursWorked () {return _hoursWorked;}

    public double earnings () {
        if (hoursWorked() <= 40) // no overtime 
            return hourlyWage() * hoursWorked(); 
        else 
            return 40 * hourlyWage() + (hoursWorked() - 40) * hourlyWage() * 1.5;
    }

    public String toString () {
        return String.format("hourly employee: %s\n%s: $%,.2f; %s: %,.2f", 
            super.toString(), "hourly wage", hourlyWage(), "hours worked", hoursWorked());
    }
}